Packages/ClientRuntime/Sources/Waiters/WaiterScheduler.swift

```diff
@@ -0,0 +1,85 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+/// Calculates the time for the next retry, and whether waiting should continue.
+/// Upon creation, the scheduler will allow an immediate request.
+/// After the first request, further requests are scheduled per the retry strategy formula
+/// published in the Smithy specification:
+/// https://awslabs.github.io/smithy/2.0/additional-specs/waiters.html#waiter-retries
+class WaiterScheduler {
+    /// The minimum delay between retries while waiting.
+    let minDelay: TimeInterval
+    /// The maximum delay between retries while waiting.
+    let maxDelay: TimeInterval
+    /// The maximum time to spend waiting for the retry to succeed, before a timeout error is thrown.
+    let maxWaitTime: TimeInterval
+
+    /// Returns the current Date.  For testing, this closure may be replaced to provide simulated time.
+    var now: () -> Date = { Date() }
+
+    /// The Date when the next request should take place after a retry.
+    var nextRequestDate = Date.distantPast
+    
+    /// The number of requests ("attempts") that have been made by this waiter.
+    private(set) var attempt = 0
```

Comment:
Nit: Can we name this `attempts` to stay consistent with the naming elsewhere in this PR?

Reply:
This violates my own conventions too... but I used `attempt` because that is the exact name of the variable in the pseudocode Smithy gives for the retry logic.  I'll change it to `attempts` because I don't think using a different name from the Smithy sample pseudocode will cause any confusion.

Reply:
ahh i see, yeah i agree